consensus/dolev_strong.go

```diff
@@ -0,0 +1,310 @@
+		case <-ticker.C:
+			log.Info("finished all DS rounds")
+			impl.Abort()
+		case <-impl.abortInstance:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+
+// SendMessage sends this message in order to reach consensus
+func (impl *DolevStrongConsensus) SendMessage(msg OpaqueMessage) error {
+	if len(msg.Data()) == 0 {
+		return nil
+	}
+	dataMsg := pb.MessageData{
+		Data:       msg.Data(),
+		AuthPubKey: impl.publicKey.Bytes(),
+	}
+	protocolMsg := pb.ConsensusMessage{
+		Msg:        &dataMsg,
+		Validators: []*pb.Validator{},
+	}
+	err := impl.signMessageAndAppend(&protocolMsg)
+	if err != nil {
+		log.Error("cannot sign message %v", protocolMsg)
+		return err
+	}
+	except := make(map[string]struct{})
+	except[impl.publicKey.String()] = struct{}{}
+	impl.sendToRemainingNodes(&protocolMsg, except)
+	return nil
+}
+
+func (dsci *DolevStrongInstance) authenticateSignature(validator crypto.PublicKey, signature []byte, data *pb.MessageData) bool {
+	dt, err := proto.Marshal(data)
+	if err != nil {
+		log.Error("error marshaling validator struct")
+		return false
+
+	}
+	out, err := validator.Verify(dt, signature)
+	if err != nil {
+		log.Error("error validating signature", signature, " errno:", err)
+		return false
+	}
+	return out
+}
+
+func (impl *DolevStrongConsensus) getRound() int32 {
+	return int32(impl.timer.Since(impl.startTime).Nanoseconds() / impl.dsConfig.PhaseTime.Nanoseconds())
+}
+
+//GetOutput Returns the message agreed upon in this instance of dolev string, returns nil if the message agreement failed
+func (dsci *DolevStrongInstance) GetOutput() []byte {
+	if dsci.abortProcessingInstance {
+		return nil
+	}
+	return dsci.value
+}
+
+// ReceiveMessage is the main receiver handler for a message received from a dolev strong instance running in the network
+func (dsci *DolevStrongInstance) ReceiveMessage(message *pb.ConsensusMessage) error{
+	//calculates round according to local clock
+	round := dsci.ds.getRound()
+	log.Info("received message in round %v num of signatures: %v", round, len(message.Validators))
+
+	if dsci.abortProcessingInstance {
+		return fmt.Errorf("message received on aborted isntance, round: %v", round)
+	}
+
+	if round > dsci.ds.dsConfig.NumOfAdverseries+1 {
+		return fmt.Errorf("round out of order: %v", round)
+	}
+
+	publicKeys, err := dsci.findAndValidateSignatures(message)
+	if err != nil {
+		return err
+	}
+	initiatorPubKey, err := crypto.NewPublicKey(message.Msg.AuthPubKey)
+	if _, ok := publicKeys[initiatorPubKey.String()]; !ok || err != nil {
+		return fmt.Errorf("initiator did not sign the message - aborting")
+
+	}
+	//check if there are enough signature to match round number
+	if len(publicKeys) < int(round) {
+		return fmt.Errorf("invalid number of signatures - not matching round number %v num of signatures %v", round, len(publicKeys))
+	}
+	//todo: what if the signatures are of nodes that are not in the layer?
+	if dsci.value != nil {
+		if res := bytes.Compare(dsci.value, message.Msg.Data); res != 0 {
```

Comment:
According to the protocol you should send the faulty message so that the other nodes will be aware that the sender is an adversary. Otherwise it is possible that a coalition of adversaries nodes will send the second value only to one honest nodes and only him will output null breaking and not reaching consensus

Reply:
not according to kumersian impl